tracer/src/Datadog.Trace/Propagators/W3CTraceContextPropagator.cs

```diff
@@ -7,125 +7,434 @@
+            if (samplingPriority.Equals("1", StringComparison.Ordinal))
             {
+                return 1;
+            }
+
+            if (samplingPriority.Equals("0", StringComparison.Ordinal))
+            {
+                return 0;
+            }
+
+            if (samplingPriority.Equals("-1", StringComparison.Ordinal))
+            {
+                return -1;
+            }
+
+            return int.TryParse(samplingPriority, out var result) ? result : null;
+        }
+#endif
+
+        public bool TryExtract<TCarrier, TCarrierGetter>(TCarrier carrier, TCarrierGetter carrierGetter, out SpanContext? spanContext)
+            where TCarrierGetter : struct, ICarrierGetter<TCarrier>
+        {
+            spanContext = null;
+
+            // get the "traceparent" header
+            var traceParentHeader = ParseUtility.ParseString(carrier, carrierGetter, TraceParentHeaderName)?.Trim();
+
+            if (string.IsNullOrEmpty(traceParentHeader) || !TryParseTraceParent(traceParentHeader!, out var traceParent))
+            {
+                // "traceparent" header is required
                 return false;
             }
 
+            var traceStateHeader = ParseUtility.ParseString(carrier, carrierGetter, TraceStateHeaderName)?.Trim();
+
+            if (string.IsNullOrEmpty(traceStateHeader) || !TryParseTraceState(traceStateHeader!, out var traceState))
+            {
+                // "tracestate" header is optional
+                traceState = default;
+            }
+
+            // if we can't get the more specific sampling priority from "tracestate",
+            // then fallback to the boolean in "traceparent"
+            var samplingPriority = traceState.SamplingPriority ??
+                                   (traceParent.Sampled ?
+                                        SamplingPriorityValues.AutoKeep :
+                                        SamplingPriorityValues.AutoReject);
+
+            spanContext = new SpanContext(
+                traceParent.TraceId,
+                traceParent.ParentId,
+                samplingPriority,
+                serviceName: null,
+                origin: traceState.Origin,
+                traceParent.RawTraceId,
+                traceParent.RawParentId);
+
+            spanContext.PropagatedTags = traceState.PropagatedTags;
             return true;
         }
 
-        private static bool IsValidSpanId([NotNullWhen(true)] string? spanId)
+        private static bool IsValidHexString([NotNullWhen(true)] string? value, int length)
         {
-            if (string.IsNullOrEmpty(spanId))
+            if (value?.Length != length)
             {
                 return false;
             }
 
-            if (spanId!.Length != 16)
+            // ReSharper disable once ForCanBeConvertedToForeach
+            for (var i = 0; i < value.Length; i++)
             {
-                return false;
+                if (value[i] is (< '0' or > '9') and (< 'a' or > 'f'))
```

Comment:
This assumes it must be lower case, is that correct? Should we accept uppercase hex?